server/src/main/java/io/crate/execution/ddl/tables/AlterTableClient.java

```diff
@@ -120,32 +139,45 @@ public AlterTableClient(ClusterService clusterService,
         this.sessions = sessions;
         this.indexScopedSettings = indexScopedSettings;
         this.logicalReplicationService = logicalReplicationService;
+        this.threadPool = threadPool;
     }
 
     public CompletableFuture<Long> addColumn(AddColumnRequest addColumnRequest) {
         var newReferences = addColumnRequest.references();
-        String subject = null;
+        String exceptionSubject = null;
+        String warning = null;
         if (addColumnRequest.pKeyIndices().isEmpty() == false) {
-            subject = "primary key";
-        } else if (addColumnRequest.references().stream().anyMatch(ref -> ref instanceof GeneratedReference)) {
-            subject = "generated";
+            exceptionSubject = "primary key";
+        } else if (newReferences.stream().anyMatch(ref -> ref instanceof GeneratedReference)) {
+            exceptionSubject = "generated";
         } else {
             for (Reference newRef : newReferences) {
                 if (newReferences.stream().anyMatch(r -> r.column().isChildOf(newRef.column()))
                     && newRef.columnPolicy().equals(ColumnPolicy.IGNORED)) {
-                    subject = "sub column to an OBJECT(IGNORED) parent";
+                    warning = "Adding a sub column to an OBJECT(IGNORED) parent may shade existing data of this column as the table isn't empty";
                     break;
                 }
             }
         }
-        if (subject != null) {
-            String finalSubject = subject;
+        if (exceptionSubject != null || warning != null) {
+            String finalSubject = exceptionSubject;
+            String finalWarning = warning;
             return getRowCount(addColumnRequest.relationName()).thenCompose(rowCount -> {
                 if (rowCount > 0) {
-                    throw new UnsupportedOperationException("Cannot add a " + finalSubject + " column to a table that isn't empty");
-                } else {
-                    return transportAddColumnAction.execute(addColumnRequest).thenApply(_ -> -1L);
+                    if (finalSubject != null) {
+                        throw new UnsupportedOperationException("Cannot add a " + finalSubject + " column to a table that isn't empty");
+                    }
+                    LOGGER.warn(finalWarning);
+                    return isGlobalCheckpointInSync(addColumnRequest.relationName())
+                        .thenCompose(isInSync -> {
+                            if (isInSync) {
+                                return transportAddColumnAction.execute(addColumnRequest).thenApply(_ -> -1L);
+                            }
+                            throw new IllegalStateException("Global checkpoint is not in sync. " +
```

Comment:
Does this mean that we essentially can't run an alter table command if there is concurrent indexing happening?  At any rate, I think we should probably refer here to replicas not being up to date with the primary, rather than to global checkpoints which aren't really things that users need to worry about, and maybe suggest pausing indexing as a solution?

Reply:
> Does this mean that we essentially can't run an alter table command if there is concurrent indexing happening? 

Yes as otherwise we may loose writes on the replica (cast exceptions on processing the translog).

Good point about the user-facing message, will adjust.